src/hooks/op_rewrite.c

```diff
@@ -0,0 +1,262 @@
+        elog(ERROR, "Failed to find the function for operator class");
+    }
+    Oid functionId = ((Form_pg_amproc)GETSTRUCT(opTuple))->amproc;
+    ReleaseSysCache(opTuple);
+
+    return functionId;
+}
+
+static Node *operator_rewriting_mutator(Node *node, void *ctx)
+{
+    OpRewriterContext *context = (OpRewriterContext *)ctx;
+
+    if(node == NULL) return node;
+
+    if(IsA(node, OpExpr)) {
+        OpExpr *opExpr = (OpExpr *)node;
+        if(list_member_oid(context->ldb_ops, opExpr->opno)) {
+            if(context->indices == NULL) {
+                return node;
+            } else {
+                ListCell *lc;
+                foreach(lc, context->indices) {
+                    uintptr_t intermediate = (uintptr_t)lfirst(lc);
+                    Oid       indexid = (Oid)intermediate;
+                    Relation  index = index_open(indexid, AccessShareLock);
+                    Oid       indexfunc = get_func_id_from_index(index);
+                    if(OidIsValid(indexfunc)) {
+                        FuncExpr *fnExpr = makeNode(FuncExpr);
+                        fnExpr->funcresulttype = opExpr->opresulttype;
+                        fnExpr->funcretset = opExpr->opretset;
+                        fnExpr->funccollid = opExpr->opcollid;
+                        fnExpr->inputcollid = opExpr->inputcollid;
+                        fnExpr->args = opExpr->args;
+                        fnExpr->location = opExpr->location;
+                        // operators can't take variadic arguments
+                        fnExpr->funcvariadic = false;
+                        // print it as a function
+                        fnExpr->funcformat = COERCE_EXPLICIT_CALL;
+                        fnExpr->funcid = indexfunc;
+
+                        index_close(index, AccessShareLock);
+
+                        return (Node *)fnExpr;
+                    }
+                    index_close(index, AccessShareLock);
+                }
+                return node;
+            }
+        }
+    }
+
+    if(IsA(node, IndexScan) || IsA(node, IndexOnlyScan)) {
+        return node;
+    }
+    if(IsA(node, SeqScan)) {
+        SeqScan *seqscan = (SeqScan *)node;
+#if PG_VERSION_NUM >= 150000
+        Plan *seqscanplan = &seqscan->scan.plan;
+        Oid   rtrelid = seqscan->scan.scanrelid;
+#else
+        Plan *seqscanplan = &seqscan->plan;
+        Oid   rtrelid = seqscan->scanrelid;
+#endif
+        RangeTblEntry *rte = rt_fetch(rtrelid, context->rtable);
+        Oid            relid = rte->relid;
+        Relation       rel = relation_open(relid, AccessShareLock);
+        if(rel->rd_indexvalid) {
+            context->indices = RelationGetIndexList(rel);
+        }
+        relation_close(rel, AccessShareLock);
+
+        base_plan_mutator(seqscanplan, context);
+        return (Node *)seqscan;
+    }
+
+    // todo:: there is a function called query_or_expression_tree_mutator that might be able to replace the custom plan
```

Comment:
My understanding is that query nodes, plan nodes, and expression nodes are all different things, with some overlap. The parser produces query nodes, and the planner (which is what this file is working with) produces plan nodes, and both work with expression nodes as well. But something like SeqScan is exclusive to the plan node type, and hence, separate handling is required.

Reply:
Thanks,  this substantially clarifies my understanding! I will remove this